File: src/sdk/typescript/test-fixtures/src/mockBuilder.ts
// src/sdk/typescript/test-fixtures/src/mockBuilder.ts
//
// k1s0 TypeScript SDK test-fixtures: Mock builder fluent API（領域 3、ADR-TEST-010 §3）。
// リリース時点で 3 service（State / Audit / PubSub）を提供。
// 採用初期で +3 (Workflow / Decision / Secret)、運用拡大時で残 6 を追加。

// Phase marker exception（採用初期で real 実装すべき builder を呼んだ時に投げる）
//
// Go の panic / Rust の FixtureError::Unimplemented / .NET の FixturePhaseUnsupportedException と対称。
// 標準 Error / TypeError ではなく ADR cite 付き独自 class で「設計上の段階展開」を明示する。
export class FixturePhaseUnsupportedError extends Error {
  constructor(
    readonly service: string,
    readonly phase: string,
  ) {
    super(`ADR-TEST-010 PHASE: ${service} mock builder は${phase}で実装（リリース時点 phase marker）`);
    this.name = 'FixturePhaseUnsupportedError';
  }
}

// 12 service の mock builder への entry point
export class MockBuilderRoot {
  constructor(private readonly defaultTenant: string) {}

  state(): StateMockBuilder {
    return new StateMockBuilder(this.defaultTenant);
  }

  audit(): AuditMockBuilder {
    return new AuditMockBuilder(this.defaultTenant);
  }

  pubsub(): PubSubMockBuilder {
    return new PubSubMockBuilder(this.defaultTenant);
  }

  // Workflow / Decision / Secret は採用初期で real 実装する phase marker
  workflow(): never {
    throw new FixturePhaseUnsupportedError('Workflow', '採用初期');
  }
  decision(): never {
    throw new FixturePhaseUnsupportedError('Decision', '採用初期');
  }
  secret(): never {
    throw new FixturePhaseUnsupportedError('Secret', '採用初期');
  }
}

// State service mock data の fluent builder
export class StateMockBuilder {
  private tenant: string;
  private key = '';
  private value: Uint8Array = new Uint8Array();
  private ttl = 0;

  constructor(tenant: string) {
    this.tenant = tenant;
  }

  withTenant(tenant: string): this { this.tenant = tenant; return this; }
  withKey(key: string): this { this.key = key; return this; }
  withValue(value: Uint8Array): this { this.value = value; return this; }
  withTtl(seconds: number): this { this.ttl = seconds; return this; }

  build(): StateEntry {
    return { tenant: this.tenant, key: this.key, value: this.value, ttl: this.ttl };
  }
}

export interface StateEntry {
  tenant: string;
  key: string;
  value: Uint8Array;
  ttl: number;
}

// Audit service mock data の fluent builder
export class AuditMockBuilder {
  private tenant: string;
  private entryCount = 0;
  private startSeq = 0;

  constructor(tenant: string) {
    this.tenant = tenant;
  }

  withTenant(tenant: string): this { this.tenant = tenant; return this; }
  withEntries(n: number): this { this.entryCount = n; return this; }
  withSequence(seq: number): this { this.startSeq = seq; return this; }

  build(): AuditEntry[] {
    const entries: AuditEntry[] = [];
    for (let i = 0; i < this.entryCount; i++) {
      entries.push({
        tenant: this.tenant,
        sequence: this.startSeq + i,
        // 採用初期で SHA-256 prev_id chain を計算
        prevId: '',
      });
    }
    return entries;
  }
}

export interface AuditEntry {
  tenant: string;
  sequence: number;
  prevId: string;
}

// PubSub service mock data の fluent builder
export class PubSubMockBuilder {
  private tenant: string;
  private topic = '';
  private messages = 0;
  private delayMs = 0;

  constructor(tenant: string) {
    this.tenant = tenant;
  }

  withTenant(tenant: string): this { this.tenant = tenant; return this; }
  withTopic(topic: string): this { this.topic = topic; return this; }
  withMessages(n: number): this { this.messages = n; return this; }
  withDelayMs(ms: number): this { this.delayMs = ms; return this; }

  build(): PubSubMessage[] {
    const msgs: PubSubMessage[] = [];
    for (let i = 0; i < this.messages; i++) {
      msgs.push({ tenant: this.tenant, topic: this.topic, seqId: i });
    }
    return msgs;
  }
}

export interface PubSubMessage {
  tenant: string;
  topic: string;
  seqId: number;
}

File: src/sdk/rust/crates/k1s0-sdk-test-fixtures/src/mock_builder.rs
//! Mock builder fluent API（領域 3、ADR-TEST-010 §3）。
//!
//! リリース時点で 3 service（State / Audit / PubSub）を提供。
//! 採用初期で +3 (Workflow / Decision / Secret)、運用拡大時で残 6 service を追加。

use crate::{FixtureError, Result};

/// 12 service の mock builder への entry point。
pub struct MockBuilderRoot {
    /// 既定 tenant（builder の WithTenant 未指定時に使う）
    default_tenant: String,
}

impl MockBuilderRoot {
    /// crate 内から呼ばれる constructor
    pub(crate) fn new(default_tenant: String) -> Self {
        Self { default_tenant }
    }

    /// State service の mock builder
    pub fn state(&self) -> StateMockBuilder {
        StateMockBuilder {
            tenant: self.default_tenant.clone(),
            key: String::new(),
            value: Vec::new(),
            ttl: 0,
        }
    }

    /// Audit service の mock builder
    pub fn audit(&self) -> AuditMockBuilder {
        AuditMockBuilder {
            tenant: self.default_tenant.clone(),
            entry_count: 0,
            start_seq: 0,
        }
    }

    /// PubSub service の mock builder
    pub fn pubsub(&self) -> PubSubMockBuilder {
        PubSubMockBuilder {
            tenant: self.default_tenant.clone(),
            topic: String::new(),
            messages: 0,
            delay_ms: 0,
        }
    }

    /// Workflow は採用初期で実装
    pub fn workflow(&self) -> Result<()> {
        Err(FixtureError::Unimplemented {
            service: "Workflow".to_string(),
            phase: "採用初期".to_string(),
        })
    }
}

// ----- State service mock builder ---------------------------------

/// State service mock data の fluent builder
pub struct StateMockBuilder {
    tenant: String,
    key: String,
    value: Vec<u8>,
    ttl: u32,
}

impl StateMockBuilder {
    /// tenant 指定（fluent chain）
    pub fn with_tenant(mut self, tenant: impl Into<String>) -> Self {
        self.tenant = tenant.into();
        self
    }
    /// key 指定
    pub fn with_key(mut self, key: impl Into<String>) -> Self {
        self.key = key.into();
        self
    }
    /// value 指定
    pub fn with_value(mut self, value: impl Into<Vec<u8>>) -> Self {
        self.value = value.into();
        self
    }
    /// TTL 秒指定
    pub fn with_ttl(mut self, seconds: u32) -> Self {
        self.ttl = seconds;
        self
    }
    /// build: 最終的な StateEntry を返す（採用初期で contracts/proto 型に置換）
    pub fn build(self) -> Result<StateEntry> {
        Ok(StateEntry {
            tenant: self.tenant,
            key: self.key,
            value: self.value,
            ttl: self.ttl,
        })
    }
}

/// State service の wire 形式（採用初期で contracts/proto 型に置換）
#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct StateEntry {
    pub tenant: String,
    pub key: String,
    pub value: Vec<u8>,
    pub ttl: u32,
}

// ----- Audit service mock builder ---------------------------------

/// Audit service mock data の fluent builder
pub struct AuditMockBuilder {
    tenant: String,
    entry_count: u32,
    start_seq: u64,
}

impl AuditMockBuilder {
    pub fn with_tenant(mut self, tenant: impl Into<String>) -> Self {
        self.tenant = tenant.into();
        self
    }
    /// entry 件数指定（hash chain で連結された N 件を生成）
    pub fn with_entries(mut self, n: u32) -> Self {
        self.entry_count = n;
        self
    }
    pub fn with_sequence(mut self, seq: u64) -> Self {
        self.start_seq = seq;
        self
    }
    /// build: AuditEntry の Vec を返す（採用初期で hash chain 計算 + proto 型に置換）
    pub fn build(self) -> Result<Vec<AuditEntry>> {
        let mut entries = Vec::with_capacity(self.entry_count as usize);
        for i in 0..self.entry_count {
            entries.push(AuditEntry {
                tenant: self.tenant.clone(),
                sequence: self.start_seq + i as u64,
                // 採用初期で prev_id chain を SHA-256 で計算
                prev_id: String::new(),
            });
        }
        Ok(entries)
    }
}

/// Audit service の wire 形式
#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct AuditEntry {
    pub tenant: String,
    pub sequence: u64,
    pub prev_id: String,
}

// ----- PubSub service mock builder --------------------------------

/// PubSub service mock data の fluent builder
pub struct PubSubMockBuilder {
    tenant: String,
    topic: String,
    messages: u32,
    delay_ms: u32,
}

impl PubSubMockBuilder {
    pub fn with_tenant(mut self, tenant: impl Into<String>) -> Self {
        self.tenant = tenant.into();
        self
    }
    pub fn with_topic(mut self, topic: impl Into<String>) -> Self {
        self.topic = topic.into();
        self
    }
    pub fn with_messages(mut self, n: u32) -> Self {
        self.messages = n;
        self
    }
    pub fn with_delay_ms(mut self, ms: u32) -> Self {
        self.delay_ms = ms;
        self
    }
    pub fn build(self) -> Result<Vec<PubSubMessage>> {
        let mut msgs = Vec::with_capacity(self.messages as usize);
        for i in 0..self.messages {
            msgs.push(PubSubMessage {
                tenant: self.tenant.clone(),
                topic: self.topic.clone(),
                seq_id: i as u64,
            });
        }
        Ok(msgs)
    }
}

/// PubSub service の wire 形式
#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct PubSubMessage {
    pub tenant: String,
    pub topic: String,
    pub seq_id: u64,
}
